libp2p/protocols/relayv2/relayv2.nim

```diff
@@ -0,0 +1,860 @@
+
+  let
+    msg = StopMessage(
+      msgType: StopMessageType.Status,
+      peer: none(RelayV2Peer),
+      limit: none(RelayV2Limit),
+      status: some(RelayV2Status.Ok))
+    pb = encodeStopMessage(msg)
+
+  try:
+    await conn.writeLp(pb.buffer)
+  except CancelledError as exc:
+    raise exc
+  except CatchableError as exc:
+    trace "error writing stop msg", exc=exc.msg
+    return
+  await cl.queue.addLast(RelayConnection.new(conn, limitDuration, limitData))
+  await conn.join()
+
+proc init*(cl: Client) =
+  proc handleStream(conn: Connection, proto: string) {.async, gcsafe.} =
+    try:
+      let msgOpt = decodeStopMessage(await conn.readLp(MsgSize))
+
+      if msgOpt.isNone():
+        await handleHopError(conn, RelayV2Status.MalformedMessage)
+        return
+      else:
+        trace "client handle stream", msg = msgOpt.get()
+      let msg = msgOpt.get()
+
+      if msg.msgType == StopMessageType.Connect:
+        await cl.handleConnect(conn, msg)
+      else:
+        trace "Unexpected client / relayv2 handshake", msgType=msg.msgType
+        await handleStopError(conn, RelayV2Status.MalformedMessage)
+    except CancelledError as exc:
+      raise exc
+    except CatchableError as exc:
+      trace "exception in client handler", exc = exc.msg, conn
+    finally:
+      trace "exiting client handler", conn
+      await conn.close()
+
+  cl.queue = newAsyncQueue[Connection](0)
+
+  cl.handler = handleStream
+  cl.codecs = @[RelayV2StopCodec]
+
+type
+  Rsvp = object
+    expire: uint64           # required, Unix expiration time (UTC)
+    addrs: seq[MultiAddress] # relay address for reserving peer
+    voucher: Option[Voucher] # optional, reservation voucher
+    limitDuration: uint32    # seconds
+    limitData: uint64        # bytes
+  ReservationError* = object of RelayV2Error
+  RelayV2DialError* = object of RelayV2Error
+
+
+proc reserve*(cl: Client, relayPI: PeerInfo): Future[Rsvp] {.async.} =
+  var
+    msg = HopMessage(msgType: HopMessageType.Reserve)
+    msgOpt: Option[HopMessage]
+  let pb = encodeHopMessage(msg)
+
+  let conn = try:
+    await cl.switch.dial(relayPI.peerId, relayPI.addrs, RelayV2HopCodec)
+  except CancelledError as exc:
+    raise exc
+  except CatchableError as exc:
+    trace "error opening relay stream", exc=exc.msg, relayPI
+    raise newException(ReservationError, exc.msg)
+  defer:
+    await conn.close()
+
+  try:
+    await conn.writeLp(pb.buffer)
+    msgOpt = decodeHopMessage(await conn.readLp(MsgSize))
+  except CancelledError as exc:
+    raise exc
+  except CatchableError as exc:
+    trace "error writing reservation message", exc=exc.msg
+    raise newException(ReservationError, exc.msg)
+
+  if msgOpt.isNone():
+    trace "Malformed message from relay"
+    raise newException(ReservationError, "Malformed message from relay")
+  msg = msgOpt.get()
+  if msg.msgType != HopMessageType.Status:
+    trace "unexpected relay response type", msgType = msg.msgType
```

Comment:
All theses trace can be removed, since the exception handler will show the error anyway